stdlib/public/core/IntegerParsing.swift

```diff
@@ -2,23 +2,166 @@
+      ? temporary.subtractingReportingOverflow(digitValue)
+      : temporary.addingReportingOverflow(digitValue)
+    guard _fastPath(!overflow2) else { return nil }
+    result = nextResult
+  }
+  return result
+}
+
+@_alwaysEmitIntoClient
+internal func _parseASCII<Result: FixedWidthInteger>(
+  _ codeUnits: UnsafeBufferPointer<UInt8>, radix: Int
+) -> Result? {
+  _internalInvariant(!codeUnits.isEmpty)
+  let first = codeUnits[0]
+  if first == 45 /* "-" */ {
+    return _parseASCIIDigits(
+      UnsafeBufferPointer(rebasing: codeUnits[1...]),
+      radix: radix, isNegative: true)
+  }
+  if first == 43 /* "+" */ {
+    return _parseASCIIDigits(
+      UnsafeBufferPointer(rebasing: codeUnits[1...]),
+      radix: radix, isNegative: false)
+  }
+  return _parseASCIIDigits(codeUnits, radix: radix, isNegative: false)
+}
+
+@_alwaysEmitIntoClient
+@inline(never)
+internal func _parseASCII<S: StringProtocol, Result: FixedWidthInteger>(
+  _ text: S, radix: Int
+) -> Result? {
+  var str = String(text)
+  return str.withUTF8 { _parseASCII($0, radix: radix) }
+}
+
+extension FixedWidthInteger {
+  /// Creates a new integer value from the given string and radix.
+  ///
+  /// The string passed as `text` may begin with a plus or minus sign character
+  /// (`+` or `-`), followed by one or more numeric digits (`0-9`) or letters
+  /// (`a-z` or `A-Z`). Parsing of the string is case insensitive.
+  ///
+  ///     let x = Int("123")
+  ///     // x == 123
+  ///
+  ///     let y = Int("-123", radix: 8)
+  ///     // y == -83
+  ///     let y = Int("+123", radix: 8)
+  ///     // y == +83
+  ///
+  ///     let z = Int("07b", radix: 16)
+  ///     // z == 123
+  ///
+  /// If `text` is in an invalid format or contains characters that are out of
+  /// bounds for the given `radix`, or if the value it denotes in the given
+  /// `radix` is not representable, the result is `nil`. For example, the
+  /// following conversions result in `nil`:
+  ///
+  ///     Int(" 100")                       // Includes whitespace
+  ///     Int("21-50")                      // Invalid format
+  ///     Int("ff6600")                     // Characters out of bounds
+  ///     Int("zzzzzzzzzzzzz", radix: 36)   // Out of range
+  ///
+  /// - Parameters:
+  ///   - text: The ASCII representation of a number in the radix passed as
+  ///     `radix`.
+  ///   - radix: The radix, or base, to use for converting `text` to an integer
+  ///     value. `radix` must be in the range `2...36`. The default is 10.
+  @inlinable
+  @inline(__always)
+  public init?<S: StringProtocol>(_ text: S, radix: Int = 10) {
+    _precondition(2...36 ~= radix, "Radix not in range 2...36")
+    guard _fastPath(!text.isEmpty) else { return nil }
+    let result: Self? =
+      text.utf8.withContiguousStorageIfAvailable {
+        _parseASCII($0, radix: radix)
+      } ?? _parseASCII(text, radix: radix)
+    guard let result_ = result else { return nil }
+    self = result_
+  }
```

Comment:
Did you compare codegen to just using `withUTF8`? Even if these are lazily-bridged Objective-C strings, there's still a good chance they'll fit in 15 code units worth of small-string.

That would allow us to avoid having to call a generic non-inlinable function, hopefully reducing or simplifying code.

Another option might be to try to concretely parse it as a wider integer (e.g. `UInt64`) and narrow it after the fact. We'd still need a generic fall-back path, but it could help if it eliminates the `FixedWidthInteger` for all non-big-int cases.

Reply:
Note that `withUTF8` is not a protocol requirement of `StringProtocol`. Calling `withUTF8` requires unconditionally creating a mutable `String`; I can try it but I'm skeptical that it'd beat the status quo.

Parsing as a wider integer could be worth trying. If I catch your meaning, you're interested to see what happens if I explicitly mark the fast path with `@_specialize(where Result == UInt64)` and then call `_parseASCII(_:radix:) -> UInt64?` here. I'd be interested to see how it goes.

Reply:
> If I catch your meaning, you're interested to see what happens if I explicitly mark the fast path with @_specialize(where Result == UInt64) and then call _parseASCII(_:radix:) -> UInt64?

Right, get rid of the generic when possible (and we may have to increase our testing coverage to test the generic path).

Reply:
I should have some benchmarks soon for this. If it pans out, we'll certainly need to increase our testing coverage.

Reply:
I wonder if this entirely problem would be better expressed in terms of a single, entirely non-generic, `@usableFromInline` entry point like this one:

```swift
/// Parse as many digits as possible from `source` (dropping them from the buffer), 
/// returning the partial digits parsed, as well as their associated scale factor.
func _parseASCIIDigits(in source: inout UnsafeBufferPointer<UInt8>, radix: Int) -> (value: UInt64, scale: UInt64)?

if Result.bitWidth <= 64 {
  // Call _parseASCIIDigits once, handling error & overflow
} else {
  var value: Result = 0
  while !buffer.isEmpty {
    guard let (nextDigits, scale) = _parseASCIIDigits(in: &buffer, radix: radix) else { return nil }
    value *= scale // except with multipliedReportingOverflow
    value += nextDigits // except with addingReportingOverflow
  }
  return value
}
```

Reply:
That's a _very_ interesting idea. I'll give that a shot when I have some time.

Reply:
(The problem of availability issues around a new `@usableFromInline` function can perhaps be worked around by this pattern, which I'm dubbing the Curiously Recursive @usableFromInline Dance, or CRUD for short:)

```swift
@available(macOS 9999, etc.)
@usableFromInline internal func _parseASCIIDigits_impl(...) {
  _parseASCIIDigits(..., forward = false)
}

@_alwaysEmitIntoClient internal func _parseASCIIDigits(..., forward: Bool = true) {
  if forward, #available(macOS 9999, etc.) {
    return _parseASCIIDigits_impl(...)
  }
  // the actual implementation
}
```

Over the years, as clients' deployment target get bumped beyond whatever 9999 becomes, the else branch in _parseASCIIDigits will gradually get constant folded away, and we get rid of the cost of emitting the core non-generic parsing code. (We'll also be able to switch the `@usableFromInline` variant to start using a purely internal implementation whenever we have a reason to do so.)

Reply:
Note: the code above isn't quite right -- UInt64-sized digit chunks would probably need to be parsed backwards in the BigInt case, or (more likely) the first chunk processed chunk sometimes needs to be partial so that the input buffer ends with a full chunk.

[This piece of code](https://github.com/attaswift/BigInt/blob/master/Sources/String%20Conversion.swift#L39) works using a similar scheme. (Although it is a bit naive in its execution -- allocating an array? not having a precomputed table of radix → (digit count, scale) mappings? pfft.)

Reply:
Note that the string-to-integer parsing here is generic over `FixedWidthInteger` specifically, so we won't have to worry about BigInt here (for now).

Reply:
Oh, FixedWidthInteger can technically still be >64 bits -- `UInt128` is definitely something we're planning to add, but things like `UInt1024`/`UInt2048` would probably have some practical use cases too. 

(It's likely best to keep as much processing as possible  in terms of Int-sized integers -- I'm pretty sure that with these larger types, doing a full mul+add over each digit would be quite a bit slower than doing it over single-word chunks.)

Reply:
Sure, we definitely do want to give some consideration for wider fixed-width integers.

I've started to poke around at what it would take to use an approach such as you suggest. It still seems promising, but it is perhaps not a slam-dunk as initially I thought.

Parsing the string backwards would be an advantage for base 16 (and for base 2 and 4), but it's not so clear for the others, since multiplying and adding each "chunk" would potentially change all words of the accumulated value when working in, say, base 10. So it's not clear to me that changing the direction of string parsing is essential here; that actually simplifies things.

The complicating factor is the following issue:

* Since `scale` is the largest power of `radix` less than (or equal to) `UInt64.max`, there are necessarily going to be values representable as a value of type `UInt64` (i.e., less than or equal to `UInt64.max`) but greater than or equal to `scale`.

* If we are converting to a type where `bitWidth <= 64`, we _want_ those values greater than or equal to `scale` returned in a single iteration of `_parseASCIIDigits`.

* If we are converting to a type in multiple chunks, we do _not_ want those values returned in any iteration of `_parseASCIIDigits`, because the corresponding `scale` would not be representable as a value of type `UInt64` *and*, even if it could (say, by scaling `scale` itself, dividing the true scale by `radix`), would vary from iteration to iteration. Instead, we want each partial value always to be less than `scale`.

The challenge is not insurmountable; I think it's still very much a feasible design. The caveat is that (unless I'm not seeing some dramatic simplification) this adds to the complexity of the code that we'll have to emit. I don't think it'll be acceptable (performance-wise) to penalize conversion of 64-bit values by potentially requiring two calls to `_parseASCIIDigits` where one would do, so either some logic bifurcating the two scenarios will need to be inserted into `_parseASCIIDigits`, or it will need to be handled caller-side (in `_parseASCII` somehow.

Anyway, this seems more than I'm going to be able to turn around in a limited amount of time; I'll polish up the implementation here and see if any availability workarounds can or need to be used. So long as none of this blocks future work, I think the incremental improvement here is worthwhile.

Reply:
> If we are converting to a type where bitWidth <= 64, we want those values greater than or equal to scale returned in a single iteration of _parseASCIIDigits.

That's true! But I think this simply means that we need a different (but still opaque) entry point for the `<= 64` case than we do for the `> 64` one. One other reason we'd need that is because we don't want the simple one-word case to spend time on calculating the `scale`, only for it to get discarded.

> If we are converting to a type in multiple chunks, we do not want those values returned in any iteration of _parseASCIIDigits, because the corresponding scale would not be representable as a value of type UInt64 and, even if it could (say, by scaling scale itself, dividing the true scale by radix), would vary from iteration to iteration. Instead, we want each partial value always to be less than scale.

Well, in BigInt, I handled the `scale == 2^64` case by assigning a special value for it. Something like that is definitely an option! Sadly I don't understand what you mean by the scale varying from iteration to iteration -- as long as the first call consumes just enough digits to align the rest so that the last iteration consumes the entire buffer, `scale` can simply be a function of the radix. (Why would a changing scale be a problem, though?)

> The caveat is that (unless I'm not seeing some dramatic simplification) this adds to the complexity of the code that we'll have to emit.

It shouldn't! The trick is that in the case we need to care about (i.e., native string (or ASCII Cocoa) with `bitWidth <= 64`), the conversion reduces to

1. A very inlinable `withUTF8` call on String/Substring. (Depending on code/compiler behavior, inlining this may create several copies of the closure body; I expect something like this may contribute to the code size explosion we're seeing. We may be able to mitigate that by tweaking withUTF8 (e.g., adding an `inline(never)` alternative), but reducing the size of the closure is also going to be a huge help.)
2. One opaque call to the single-shot _parseASCIIDigits entry point. 
    This will need to go through CRUD, but the code size impact of that can be mitigated by marking the @_aeic entry point `@inline(never)`. (The entry point is not generic, so it doesn't need to be specialized -- so there will only ever be a single copy emitted to each client. (Ignoring compiler weirdness like dead argument specializations, which are a mystery to me. 🙈))
    If the min.depl.target is set high enough (which it will almost always be in a couple years), the emitted CRUD function will just trivially forward to the opaque one.
3. One narrowing integer conversion on the result.
4. A couple of branches for handling failure cases (expected to be very predictable)

In the `bitWidth > 64` case, I expect switching to word-based parsing will mostly be a time rather than space improvement (by greatly reducing the number of wide multiplications), although the tweaks we apply to the single shot case will probably help code size for these, too!

It is very much possible you're seeing further than I do and see complications I don't! In any case, I think I have talked myself into trying to implement this. 😉 

I'll not be able to do it in the next few days though, and I don't think this should be holding up this PR.

Reply:
> But I think this simply means that we need a different (but still opaque) entry point for the `<= 64` case than we do for the `> 64` one.

Right, that’s what this means, and I worry that the code size issue comes from that; the remainder of the complications I mention have to do with trying _not_ to have two entry points but still keeping the fast path fast.

> Well, in BigInt, I handled the `scale == 2^64` case by assigning a special value for it.

A generalizable way to handle this for every base is to return `radix ** (digitsConsumed - 1)` instead of `radix ** digitsConsumed`, since the case where no digits are consumed can be represented by `nil`.